Sources/Workspace/DefaultPluginScriptRunner.swift

```diff
@@ -125,204 +127,257 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner, Cancellable {
+                cached: true)
+            return callbackQueue.async {
+                completion(.success(result))
+            }
+        }
 
-            // If we already have a compiled executable, then compare its hash with the new one.
-            var compilationNeeded = true
-            let hashFile = executableFile.parentDirectory.appending(component: execName + ".inputhash")
-            if fileSystem.exists(executableFile) && fileSystem.exists(hashFile) {
+        // Otherwise we need to recompile. We start by cleaning up any old files to avoid confusion if the compiler can't be invoked.
+        do {
+            try fileSystem.removeFileTree(execFilePath)
+            try fileSystem.removeFileTree(diagFilePath)
+            try fileSystem.removeFileTree(stateFilePath)
+        }
+        catch {
+            observabilityScope.emit(debug: "... couldn't clean up before invoking compiler (\(error))")
+        }
+        
+        // Now invoke the compiler asynchronously.
+        let startTime = DispatchTime.now()
+        Process.popen(arguments: commandLine, environment: toolchain.swiftCompilerEnvironment, queue: callbackQueue) {
+            // We are now on our caller's requested callback queue, so we just call the completion handler directly.
+            dispatchPrecondition(condition: .onQueue(callbackQueue))
+            completion($0.tryMap { process in
+                // Emit the compiler output as observable info.
+                let compilerOutput = ((try? process.utf8Output()) ?? "") + ((try? process.utf8stderrOutput()) ?? "")
+                observabilityScope.emit(info: compilerOutput)
+
+                // Save the persisted compilation state for possible reuse next time.
+                let compilationState = PersistedCompilationState(
+                    commandLine: commandLine,
+                    environment: toolchain.swiftCompilerEnvironment,
+                    inputHash: compilerInputHash,
+                    duration: Double(startTime.distance(to: .now()).milliseconds() ?? 0) / 1000.0,
+                    output: compilerOutput,
+                    result: .init(process.exitStatus))
                 do {
-                    if (try fileSystem.readFileContents(hashFile)) == compilerInputsHash {
-                        compilationNeeded = false
-                    }
+                    try JSONEncoder.makeWithDefaults().encode(path: stateFilePath, fileSystem: self.fileSystem, compilationState)
                 }
                 catch {
-                    // We failed to read the `.inputhash` file. We warn about it but proceed with the compilation (a cache miss).
-                    observabilityScope.emit(warning: "Couldn't read previous hash of plugin compilation inputs (\(error)")
+                    // We couldn't write out the `.state` file. We warn about it but proceed.
+                    observabilityScope.emit(debug: "... couldn't save plugin compilation state (\(error))")
```

Comment:
curious about the "..." prefix in these debug messages

Reply:
Probably just left over from print-style debugging (my indication for the end of an activity, whereas the beginning has a "..." suffix instead).  I'll changing it.